sql_metadata/parser.py

```diff
@@ -0,0 +1,533 @@
+"""
+This module provides SQL query parsing functions
+"""
+import re
+from typing import Dict, List, Optional, Tuple
+
+import sqlparse
+from sqlparse.sql import TokenList
+from sqlparse.tokens import Name, Number, Punctuation, Whitespace, Wildcard
+
+from sql_metadata.generalizator import Generalizator
+from sql_metadata.keywords_lists import (
+    FUNCTIONS_IGNORED,
+    KEYWORDS_BEFORE_COLUMNS,
+    KEYWORDS_IGNORED,
+    SUBQUERY_PRECEDING_KEYWORDS,
+    TABLE_ADJUSTMENT_KEYWORDS,
+    WITH_ENDING_KEYWORDS,
+)
+from sql_metadata.token import EmptyToken, SQLToken
+from sql_metadata.utils import UniqueList
+
+
+class Parser:  # pylint: disable=R0902
+    """
+    Main class to parse sql query
+    """
+
+    def __init__(self, sql: Optional[str] = None) -> None:
+        self._raw_query = sql
+        self._query = self._preprocess_query()
+
+        self._tokens = None
+
+        self._columns = None
+        self._columns_dict = None
+
+        self._tables = None
+        self._table_aliases = None
+
+        self._with_names = None
+        self._subqueries = None
+        self._subqueries_names = None
+
+        self._limit_and_offset = None
+
+        self._values = None
+        self._values_dict = None
+
+    @property
+    def query(self) -> str:
+        """
+        Returns preprocessed query
+        """
+        return self._query
+
+    @query.setter
+    def query(self, value):
+        """
+        Resets cached attributes before setting new query
+        """
+        self._tokens = None
+
+        self._columns = None
+        self._columns_dict = None
+
+        self._tables = None
+        self._table_aliases = None
+
+        self._with_names = None
+        self._subqueries = None
+        self._subqueries_names = None
+
+        self._limit_and_offset = None
+
+        self._values = None
+        self._values_dict = None
+
+        self._raw_query = value
```

Comment:
I'm wondering if this is needed. Is there any scenario where `self._query` will change? I believe `_query` is only set (and pre-processed) when `Parser` instance is created.

Reply:
You still can manually assign Parser.query = "New sql" and either we need to raise exception (not allowed, use new instance or something), or reset the cached props as they will still have values from old query and will return them instead of running

Reply:
I'd rather not allow people to do that. Let's assume that `Parser` instance is run for a specific query. Want to parse another query - create a new instance....

Reply:
I think just removing query.setter should do the trick

Reply:
Agree. Let's do that 👌🏻

Reply:
Removed.